src/core/shared_memory_manager.cc

```diff
@@ -117,34 +151,73 @@ SharedMemoryManager::RegisterSharedMemory(
   }
 
   // register (or re-register)
-  void* mapped_addr;
-  int shm_fd = -1;
-
-  // don't re-open if shared memory is already open
-  for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
-       ++itr) {
-    if (itr->second->shm_key_ == shm_key) {
-      shm_fd = itr->second->shm_fd_;
-      break;
+  if (kind == 0) {
+    void* mapped_addr;
+    int shm_fd = -1;
+
+    // don't re-open if shared memory is already open
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key) {
+        shm_fd = itr->second->shm_fd_;
+        break;
+      }
     }
-  }
 
-  // open and set new shm_fd if new shared memory key
-  if (shm_fd == -1) {
-    RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
-  }
+    // open and set new shm_fd if new shared memory key
+    if (shm_fd == -1) {
+      RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
+    }
+
+    Status status = MapSharedMemory(shm_fd, offset, byte_size, &mapped_addr);
+    if (!status.IsOk()) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "failed to register shared memory region '" + name + "'");
+    }
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, shm_fd, mapped_addr,
+                  nullptr, kind, device_id))));
+  } else {
+#ifdef TRTIS_ENABLE_GPU
+    void *mapped_addr = nullptr, *cuda_ipc_addr;
+    size_t shm_byte_size = 0, shm_offset = 0;
+
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key) {
+        mapped_addr = itr->second->mapped_addr_;
+        shm_byte_size = itr->second->byte_size_;
+        shm_offset = itr->second->offset_;
+        break;
```

Comment:
So the user will need to call register system shared memory first and then calls register CUDA shared memory on that region? Otherwise these variables will never get set to meaningful values.

Reply:
Yes, he must register the shared memory region that will contain the CUDA IPC handle before the CUDA IPC handle itself, if that region does not exist then. I will make sure to throw and error in that case.

Reply:
No longer relevant due to change in method of registering CudaSharedMemory